lucene/core/src/java/org/apache/lucene/codecs/blocktree/FieldReader.java

```diff
@@ -88,10 +97,13 @@
 
     if (indexIn != null) {
       final IndexInput clone = indexIn.clone();
+      //System.out.println("start=" + indexStartFP + " field=" + fieldInfo.name);
       clone.seek(indexStartFP);
-      // Initialize FST offheap if index is MMapDirectory and
-      // docCount != sumDocFreq implying field is not primary key
-      if (clone instanceof ByteBufferIndexInput && this.docCount != this.sumDocFreq) {
+
+      final boolean fstOffHeap = readerAttributes != null && "true".equals(readerAttributes.get(FST_OFFHEAP_ATTRIBUTE));
```

Comment:
don't we want to have a per-field setting? I mean something like:

`"true".equals(readerAttributes.get("blocktree.fst." + fieldInfo.name + ".offheap"));`

this would be more flexible no?

Reply:
That was the plan initially, but given that non-PK eligible fields are offheap by default. Having single setting instead of per field setting makes more sense.
```
If false:
Only non-PK fields offheap
If true:
All fields offheap
```

Reply:
why is this? I mean there might be fields that can easily be off-heap while another one should be on-heap?

Reply:
>why is this? I mean there might be fields that can easily be off-heap while another one should be on-heap?

Having per field setting is unnecessary cognitive load for the user, given that non PK eligible fields are offheap by default. So, users want PK eligible fields either offheap or onheap.
Initially, I did try it with [this patch](https://issues.apache.org/jira/secure/attachment/12958117/offheap_generic_settings.patch), but found it to be unnecessary. There can be dynamic fields as well, for which you need special keyword, which needs to account for all combinations of offheap and onheap.